### PR TITLE
Proxy CRUD functionality, isolate description service in proxy client

### PR DESCRIPTION
**Pull request by samasno**
## Trello
https://trello.com/c/aj45KkKq/12-refactor-proxy-to-forward-requests-and-split-prod-dev-environments

## Major additions
`/controllers/forwardingController.js` - implemented logic to forward request from proxy to local description service
`/public/index.html` - isolated description service on proxy page to exclude non-functioning services, description bundle served from AWS S3
`/routers/description.js` - implemented forwardingController in endpoints

## Purpose of PR
Properly forward all requests to proxy to local description service and ensure proxy connects and serves client served from s3. All crud of description service is properly served through proxy

### CODE REVIEW

File: routers/description.js [hunk 1]
```diff
@@ -5,13 +5,13 @@ const router = require('express').Router();
...
+  .get(asyncHandler(forwardingController('DESCRIPTION', 4004)))
+  .put(asyncHandler(forwardingController('DESCRIPTION', 4004)))
+  .delete(asyncHandler(forwardingController('DESCRIPTION', 4004)));
 router.route('/descriptions/new')
-  .post(asyncHandler(forwardingController('remote', 'local')));
+  .post(asyncHandler(forwardingController('DESCRIPTION', 4004)));
 
 router.route('/descriptions/multiple')
-  .get(asyncHandler(forwardingController('remote', 'local')));
+  .get(asyncHandler(forwardingController('DESCRIPTION', 4004)));
```
Comment (tyler-thee-creator) [inline routers/description.js:15]: Really cool how you've set your routes up here, love how requests filter through it and different URLs correspond with what you're doing (get single product, get multiple, POST new, etc)

File: controllers/forwarding.js [hunk 2]
```diff
@@ -1,9 +1,23 @@
...
-  //TODO handle forwarding, body, headers, routes
-  console.log('forwarding controller')
-  res.json(`${req.method}@${req.originalUrl} works. remote: ${ remote } local: ${ local }`); // test to make sure it works
-}
\ No newline at end of file
+exports.forwardingController = (envVarName = 'ENV var for service', portNumber ='port') => async (req, res, next) => {
+  // check environment, if prod use remote url env var, else use localhost port to make base of url
+  const serviceUrl = process.env.NODE_ENV === 'production' ? process.env[envVarName] : `http://localhost:${portNumber}`;
+
+  let { originalUrl, body, method, headers } = req;
```
Comment (tyler-thee-creator) [inline controllers/forwarding.js:9]: Super clean destructuring, nice

File: controllers/forwarding.js [hunk 3]
```diff
@@ -1,9 +1,23 @@
...
+
+  // seperate get from other requests, make req forwarding body and headers and return response to client
+  let response;
+  if (method === 'get') {
+    response = await axios.get(url, {headers});
+  } else {
+    response = await axios[method](url, body, {headers});
+  }
+
+  res.send(response.data);
```
Comment (tyler-thee-creator) [inline controllers/forwarding.js:22]: One potential improvement here would be to destructure right when you're making your axios call:

**Meaning:**

_let response;
if (method === 'get') {
  response = await axios.get(url, {headers});
} else {
  response = await axios[method](url, body, {headers});
}_

res.send(response.data);

**Becomes:**

_if (method === 'get') {
  const { data } = await axios.get(url, {headers});
  res.send(data);
} else {
  const { data } = await axios[method](url, body, {headers});
  res.send(data);
}_

Really doesn't matter though and comes down to preference

File: controllers/forwarding.js [hunk 4]
```diff
@@ -1,9 +1,23 @@
...
-// takes a remote url for foreign service for deployment and a local port in development
+// takes a string for the env var name for deployment and a portNumber in development
 // returns endpoint that forwards request to either remote or local args
-exports.forwardingController = (remote = 'forward requres here', local ='fallback port if local') => (req, res, next) => {
-  //TODO handle forwarding, body, headers, routes
-  console.log('forwarding controller')
-  res.json(`${req.method}@${req.originalUrl} works. remote: ${ remote } local: ${ local }`); // test to make sure it works
-}
\ No newline at end of file
+exports.forwardingController = (envVarName = 'ENV var for service', portNumber ='port') => async (req, res, next) => {
```
Comment (tyler-thee-creator) [inline controllers/forwarding.js:5]: Never seen a double assigning to a function like this:

_const variable_name = (params) => (more_params) => {
  // code
}_

Pretty cool, I'll have to look it up/check it out

**Review by tyler-thee-creator — APPROVED** [review]
Looks solid, nicely done

File: controllers/forwarding.js (hunk 3 above)
Comment (samasno) [reply controllers/forwarding.js:22]: good call. i just added the .data in as a fix when i realized i was returning the axios metadata. will correct in the next pr

File: controllers/forwarding.js (hunk 4 above)
Comment (samasno) [reply controllers/forwarding.js:5]: comes in handy. the higher order functions and currying are a game changer

**Pull request merged**